akka-stream/src/main/scala/akka/stream/impl/Timers.scala

```diff
@@ -31,88 +32,131 @@ private[stream] object Timers {
+    override def initialAttributes = DefaultAttributes.completion
+
+    override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+      new TimerGraphStageLogic(shape) with InHandler with OutHandler {
+        setHandlers(in, out, this)
 
-    override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new TimerGraphStageLogic(shape) {
-      setHandler(in, new InHandler {
         override def onPush(): Unit = push(out, grab(in))
-      })
 
-      setHandler(out, new OutHandler {
         override def onPull(): Unit = pull(in)
-      })
 
-      final override protected def onTimer(key: Any): Unit =
-        failStage(new TimeoutException(s"The stream has not been completed in $timeout."))
+        final override protected def onTimer(key: Any): Unit =
+          failStage(new TimeoutException(s"The stream has not been completed in $timeout."))
 
-      override def preStart(): Unit = scheduleOnce("CompletionTimeoutTimer", timeout)
-    }
+        override def preStart(): Unit = scheduleOnce(CompletionTimeoutTimer, timeout)
+      }
 
     override def toString = "CompletionTimeout"
+
+    case object CompletionTimeoutTimer
+
   }
 
   final class Idle[T](timeout: FiniteDuration) extends SimpleLinearGraphStage[T] {
+    override def initialAttributes = DefaultAttributes.idle
+
+    override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+      new TimerGraphStageLogic(shape) with InHandler with OutHandler {
+        private var nextDeadline: Long = System.nanoTime + timeout.toNanos
+
+        setHandlers(in, out, this)
 
-    override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new TimerGraphStageLogic(shape) {
-      private var nextDeadline: Deadline = Deadline.now + timeout
-      setHandler(in, new InHandler {
         override def onPush(): Unit = {
-          nextDeadline = Deadline.now + timeout
+          nextDeadline = System.nanoTime + timeout.toNanos
           push(out, grab(in))
         }
-      })
 
-      setHandler(out, new OutHandler {
         override def onPull(): Unit = pull(in)
-      })
 
-      final override protected def onTimer(key: Any): Unit =
-        if (nextDeadline.isOverdue())
-          failStage(new TimeoutException(s"No elements passed in the last $timeout."))
+        final override protected def onTimer(key: Any): Unit =
+          if (nextDeadline - System.nanoTime < 0)
+            failStage(new TimeoutException(s"No elements passed in the last $timeout."))
 
-      override def preStart(): Unit = schedulePeriodically("IdleTimeoutCheckTimer", interval = idleTimeoutCheckInterval(timeout))
-    }
+        override def preStart(): Unit = schedulePeriodically(IdleTimeoutCheckTimer, idleTimeoutCheckInterval(timeout))
+      }
 
     override def toString = "IdleTimeout"
+
+    case object IdleTimeoutCheckTimer
+
+  }
+
+  final class BackpressureTimeout[T](timeout: FiniteDuration) extends SimpleLinearGraphStage[T] {
```

Comment:
Demand is an internal implementation detail. Why is "backpressure timeout" bad? It is a timeout on how long you can be backpressured.